src/BSON/functions.c

```diff
@@ -0,0 +1,183 @@
+   Returns the PHP representation of a BSON value, optionally converting it into a custom class */
+PHP_FUNCTION(MongoDB_BSON_toPHP)
+{
+	char                  *data;
+	phongo_zpp_char_len    data_len;
+	zval                  *typemap = NULL;
+	php_phongo_bson_state  state = PHONGO_BSON_STATE_INITIALIZER;
+
+	SUPPRESS_UNUSED_WARNING(return_value_ptr) SUPPRESS_UNUSED_WARNING(this_ptr) SUPPRESS_UNUSED_WARNING(return_value_used) /* We don't use these */
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s|a!", &data, &data_len, &typemap) == FAILURE) {
+		return;
+	}
+
+	if (!php_phongo_bson_typemap_to_state(typemap, &state.map TSRMLS_CC)) {
+		return;
+	}
+
+	if (!php_phongo_bson_to_zval_ex((const unsigned char *)data, data_len, &state)) {
+		zval_ptr_dtor(&state.zchild);
+		RETURN_NULL();
+	}
+
+#if PHP_VERSION_ID >= 70000
+	RETURN_ZVAL(&state.zchild, 0, 1);
+#else
+	RETURN_ZVAL(state.zchild, 0, 1);
+#endif
+} /* }}} */
+
+/* {{{ proto string MongoDB\BSON\fromJSON(string $json)
+   Returns the BSON representation of a JSON value */
+PHP_FUNCTION(MongoDB_BSON_fromJSON)
+{
+	char                *json;
+	phongo_zpp_char_len  json_len;
+	bson_t               bson = BSON_INITIALIZER;
+	bson_error_t         error;
+
+	SUPPRESS_UNUSED_WARNING(return_value_ptr) SUPPRESS_UNUSED_WARNING(this_ptr) SUPPRESS_UNUSED_WARNING(return_value_used) /* We don't use these */
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s", &json, &json_len) == FAILURE) {
+		return;
+	}
+
+	if (bson_init_from_json(&bson, (const char *)json, json_len, &error)) {
+		PHONGO_RETVAL_STRINGL((const char *) bson_get_data(&bson), bson.len);
+		bson_destroy(&bson);
+	} else {
+		phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE TSRMLS_CC, "%s", error.domain == BSON_ERROR_JSON ? error.message : "Error parsing JSON");
+	}
+} /* }}} */
+
+static void phongo_bson_to_json(INTERNAL_FUNCTION_PARAMETERS, php_phongo_json_mode_t mode)
+{
+	char                *data;
+	phongo_zpp_char_len  data_len;
+	const bson_t        *bson;
+	bool                 eof = false;
+	bson_reader_t       *reader;
+	char                *json;
+	size_t               json_len;
+
+	SUPPRESS_UNUSED_WARNING(return_value_ptr) SUPPRESS_UNUSED_WARNING(this_ptr) SUPPRESS_UNUSED_WARNING(return_value_used) /* We don't use these */
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s", &data, &data_len) == FAILURE) {
+		return;
+	}
+
+	reader = bson_reader_new_from_data((const unsigned char *)data, data_len);
+	bson = bson_reader_read(reader, NULL);
+
+	if (!bson) {
+		phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE TSRMLS_CC, "Could not read document from BSON reader");
+		bson_reader_destroy(reader);
+		return;
+	}
+
+	if (mode == PHONGO_JSON_MODE_LEGACY) {
+		json = bson_as_json(bson, &json_len);
+	} else if (mode == PHONGO_JSON_MODE_CANONICAL) {
+		json = bson_as_canonical_json(bson, &json_len);
+	} else if (mode == PHONGO_JSON_MODE_RELAXED) {
+		json = bson_as_relaxed_json(bson, &json_len);
+	}
```

Comment:
The functions in this file were mostly copied as-is from `src/bson.c`. This is the only significant change, as I've added a mode enum and support for `bson_as_relaxed_json()`.